front/scripts/i18n-checker.ts

```diff
@@ -166,98 +167,174 @@ async function getLocalesKeys(localePath: string, locale: string): Promise<Set<s
-      console.warn(unusedKeys.map((e) => `${e.locale}:${e.key}`).join('\n'));
-      console.warn();
+  if (!key.includes(':')) {
+    if (prefix) {
+      key = `${prefix}.${key}`;
     }
+    key = `${defaultNamespace}:${key}`;
+  }
+
+  extractedKeys.add(key.replace(/_(zero|one|other|many)$/, ''));
+}
+
+function visitNode(
+  checker: ts.TypeChecker,
+  extractedKeys: Set<string>,
+  file: ts.SourceFile,
+  node: ts.Node
+) {
+  if (ts.isCallExpression(node)) {
+    visitCallExpression(checker, extractedKeys, file, node);
+  }
+
+  node.forEachChild((child) => visitNode(checker, extractedKeys, file, child));
+}
+
+function extractKeysFromTypeScript() {
+  const tsconfigPath = ts.findConfigFile(process.cwd(), ts.sys.fileExists, 'tsconfig.json');
+  if (!tsconfigPath) {
+    throw new Error('Failed to find tsconfig.json');
+  }
+  const tsconfigFile = ts.readConfigFile(tsconfigPath, ts.sys.readFile);
+  const tsconfig = ts.parseJsonConfigFileContent(
+    tsconfigFile.config,
+    ts.sys,
+    path.dirname(tsconfigPath)
+  );
+
+  const program = ts.createProgram({
+    options: tsconfig.options,
+    rootNames: tsconfig.fileNames,
+    projectReferences: tsconfig.projectReferences,
+  });
+  const checker = program.getTypeChecker();
 
-    if (missingKeys.length > 0) {
-      console.warn(`Missing keys (${missingKeys.length})`);
-      console.warn('------------------------------------');
-      console.warn(missingKeys.map((e) => `${e.locale}:${e.key}`).join('\n'));
-      console.warn();
-      console.warn('/!\\ Failed: missing keys are not allowed in fr & en');
-      process.exit(1);
+  const files = program.getSourceFiles();
+  const extractedKeys = new Set<string>();
+  for (const file of files) {
+    visitNode(checker, extractedKeys, file, file);
+  }
+
+  return extractedKeys;
+}
+
+const keysByLocale = await Promise.all(
+  LANGUAGES.map(async (locale) => {
+    const keys = await getLocalesKeys(`${process.cwd()}/public/locales`, locale);
+    return { locale, keys };
+  })
+);
+
+const extractedKeys = extractKeysFromTypeScript();
+
+const unusedKeys: string[] = [];
+keysByLocale.forEach(({ locale, keys }) => {
+  keys.forEach((key) => {
+    if (!extractedKeys.has(key) && IGNORE_UNUSED.every((pattern) => !key.match(pattern))) {
+      unusedKeys.push(`${locale}:${key}`);
     }
-  } catch (err) {
-    console.error(err);
-    process.exit(1);
-  } finally {
-    process.exit();
+  });
+});
+
+const missingKeys: string[] = [];
+extractedKeys.forEach((key) => {
+  if (IGNORE_MISSING.every((pattern) => !key.match(pattern))) {
+    keysByLocale.forEach(({ locale, keys }) => {
+      if (!keys.has(key)) {
+        missingKeys.push(`${locale}:${key}`);
+      }
+    });
   }
+});
+
+if (unusedKeys.length > 0) {
+  console.warn(`Unused keys (${unusedKeys.length})`);
+  console.warn('----------------------------------');
+  console.warn(unusedKeys.join('\n'));
+  console.warn();
 }
 
-run();
```

Comment:
You did not like the main function run ?

Reply:
No, I think it's unnecessary boilerplate.